packages/block-library/src/template-part/edit/template-part-previews.js

```diff
@@ -0,0 +1,209 @@
+					<PanelGroup
+						key={ templatePartList[ 0 ].meta.theme }
+						title={ templatePartList[ 0 ].meta.theme }
+					>
+						{ templatePartList.map( ( templatePart ) => {
+							return currentShownTPs.includes( templatePart ) ? (
+								<TemplatePartItem
+									key={ templatePart.id }
+									templatePart={ templatePart }
+									selectTemplate={ selectTemplate }
+								/>
+							) : (
+								<TemplatePartPlaceholder
+									key={ templatePart.id }
+								/>
+							);
+						} ) }
+					</PanelGroup>
+				) ) }
+		</>
+	);
+}
+
+function TemplatePartSearchResults( {
+	templateParts,
+	selectTemplate,
+	filterValue,
+} ) {
+	const filteredTPs = useMemo( () => {
+		// Filter based on value.
+		const lowerFilterValue = filterValue.toLowerCase();
+		const searchResults = templateParts.filter(
+			( { slug, meta: { theme } } ) =>
+				slug.toLowerCase().includes( lowerFilterValue ) ||
+				theme.toLowerCase().includes( lowerFilterValue )
+		);
+		// Order based on value location.
+		searchResults.sort( ( a, b ) => {
+			// First prioritize index found in slug.
+			const indexInSlugA = a.slug
+				.toLowerCase()
+				.indexOf( lowerFilterValue );
+			const indexInSlugB = b.slug
+				.toLowerCase()
+				.indexOf( lowerFilterValue );
+			if ( indexInSlugA !== -1 && indexInSlugB !== -1 ) {
+				return indexInSlugA - indexInSlugB;
+			} else if ( indexInSlugA !== -1 ) {
+				return -1;
+			} else if ( indexInSlugB !== -1 ) {
+				return 1;
+			}
+			// Second prioritize index found in theme.
+			return (
+				a.meta.theme.toLowerCase().indexOf( lowerFilterValue ) -
+				b.meta.theme.toLowerCase().indexOf( lowerFilterValue )
+			);
+		} );
+		return searchResults;
+	}, [ filterValue, templateParts ] );
+
+	const currentShownTPs = useAsyncList( filteredTPs );
+
+	return (
+		<>
+			{ filteredTPs.map( ( templatePart ) => (
+				<PanelGroup
+					key={ templatePart.id }
+					title={ templatePart.meta.theme }
+				>
+					{ currentShownTPs.includes( templatePart ) ? (
+						<TemplatePartItem
+							key={ templatePart.id }
+							templatePart={ templatePart }
+							selectTemplate={ selectTemplate }
+						/>
+					) : (
+						<TemplatePartPlaceholder key={ templatePart.id } />
+					) }
+				</PanelGroup>
+			) ) }
+		</>
+	);
+}
+
+export default function TemplateParts( { selectTemplate, filterValue } ) {
+	const templateParts = useSelect( ( select ) => {
+		return select( 'core' ).getEntityRecords(
+			'postType',
+			'wp_template_part',
+			{
+				status: [ 'publish', 'auto-draft' ],
+			}
+		);
+	}, [] );
+
+	if ( ! templateParts || ! templateParts.length ) {
+		return null;
+	}
+
+	if ( filterValue ) {
+		return (
+			<TemplatePartSearchResults
+				templateParts={ templateParts }
+				selectTemplate={ selectTemplate }
+				filterValue={ filterValue }
+			/>
+		);
+	}
+
+	return (
+		<TemplatePartsByTheme
+			templateParts={ templateParts }
+			selectTemplate={ selectTemplate }
```

Comment:
This duplication is unnecessary. `TemplatePartSearchResults` could be `TemplatePartsByTheme` with a single "Results" theme group.

Reply:
That makes sense 👍